src/search/pdbs/pattern_collection_generator_multiple_cegar.cc

```diff
@@ -0,0 +1,300 @@
+                             << non_goal_variables.size()
+                             << " non-goal variables: ";
+                for (int var : blacklisted_variables) {
+                    utils::g_log << var << ", ";
+                }
+                utils::g_log << endl;
+            }
+        }
+
+        int remaining_pdb_size_for_cegar = min(remaining_collection_size, max_pdb_size);
+        double remaining_time_for_cegar = min(timer.get_remaining_time().operator double(), cegar_max_time);
+        /*
+          Call CEGAR with the remaining size budget (limiting one of pdb and
+          collection size would be enough, but this is cleaner), with the
+          remaining time limit and an RNG instance with a different random
+          seed in each iteration.
+        */
+        PatternCollectionInformation collection_info = CEGAR(
+            remaining_pdb_size_for_cegar,
+            remaining_collection_size,
+            wildcard_plans,
+            remaining_time_for_cegar,
+            cegar_verbosity,
+            make_shared<utils::RandomNumberGenerator>(
+                random_seed + num_iterations),
+            task,
+            {goals[goal_index]},
+            move(blacklisted_variables)).compute_pattern_collection();
+        shared_ptr<PatternCollection> new_patterns = collection_info.get_patterns();
+        shared_ptr<PDBCollection> new_pdbs = collection_info.get_pdbs();
+        if (new_patterns->size() > 1) {
+            cerr << "a generator computed more than one pattern" << endl;
+            utils::exit_with(utils::ExitCode::SEARCH_CRITICAL_ERROR);
+        }
+
+        const Pattern &pattern = new_patterns->front();
+        if (verbosity >= utils::Verbosity::DEBUG) {
+            utils::g_log << "generated patterns " << pattern << endl;
+        }
+        if (generated_patterns.insert(pattern).second) {
+            // CEGAR generated a new pattern. Reset stagnation_start_time.
+            stagnation_start_time = -1;
+
+            shared_ptr<PatternDatabase> &pdb = new_pdbs->front();
+            remaining_collection_size -= pdb->get_size();
+            generated_pdbs->push_back(pdb);
+
+            if (remaining_collection_size <= 0) {
+                /*
+                  This value can become negative if the given size limits for
+                  pdb or collection size are so low that even the singleton
+                  goal pattern generated by CEGAR violates it.
+                */
+                if (verbosity >= utils::Verbosity::NORMAL) {
+                    utils::g_log << "collection size limit reached" << endl;
+                }
+                break;
+            }
+        } else {
+            // Pattern is not new. Set stagnation start time if not already set.
+            if (stagnation_start_time == -1) {
+                stagnation_start_time = timer.get_elapsed_time();
+            }
+        }
+
+        if (timer.is_expired()) {
```

Comment:
I'd move the code from here to line 213 to a function `bool more_iterations(...)` and replace the `while(true)` in line 88 with `while(more_iterations(...))`

Reply:
I would turn ```bool blacklisting``` and ```double stagnation_start_time``` into member variables of the class, but I'm also available on zoom if you wish to discuss this in person.